manuals/olecfmount.1
.Dd September  4, 2026
.Dt OLECFMOUNT 1
.Os
.Sh NAME
.Nm olecfmount
.Nd mounts an Object Linking and Embedding (OLE) Compound File (CF)
.Sh SYNOPSIS
.Nm olecfmount
.Op Fl c Ar codepage
.Op Fl X Ar extended_options
.Op Fl hvV
.Ar source
.Ar mount_point
.Sh DESCRIPTION
.Nm olecfmount
is a utility to mount an Object Linking and Embedding (OLE) Compound File (CF)
.Pp
.Nm olecfmount
is part of the
.Nm libolecf
package.
.Nm libolecf
is a library to access the Object Linking and Embedding (OLE) Compound File \
(CF) format
.Pp
.Ar source
an Object Linking and Embedding (OLE) Compound File (CF)
.Ar mount_point
the directory to serve as mount point
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c Ar codepage
codepage of ASCII strings, options: ascii, windows-874, windows-932, \
windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 \
(default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 \
or windows-1258
.It Fl h
shows this help
.It Fl v
verbose output to stderr, while olecfmount will remain running in the foreground
.It Fl V
print version
.It Fl X Ar extended_options
extended options to pass to sub system
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# olecfmount Text.doc /mnt/olecf
olecfmount 20260904
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr olecfexport 1 ,
.Xr olecfinfo 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libolecf/issues
.Sh COPYRIGHT
Copyright (C) 2008-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* add support for variant type and multi variant type
* double check format bounds checks
* add API support for property types
  - property values allow to set ascii codepage
  - add multi values support e.g. 0x101e
* use libfvalue table for property values in a property set section?
* fix: info_handle_document_summary_information_fprint: unable to read property set stream
* libolecf:
  - restrict stream functions to stream directory types
  - add stream read random
  - libolecf_stream_read_buffer change uint8_t* to void* ?
  - fill unallocated_block_list
* pyolecf:
  - add format version function
  - property values allow to set ascii codepage
  - pyolecf_property_value add functions
  - fix pyolecf_property_value get data function
  - add more definitions types
* olecfexport
  - add stream name to export failed error, if possible
* olecfmount
  - sanitize filenames
* support files with an empty root directory entry
* update man pages
  - correct olecfinfo.1
  - correct olecfexport.1
  - add olecfmount.1
* add support for multi types (word2003 test file)
* refactor libolecf_io_handle_read_stream into libolecf_io_handle_read_stream ?
* use libfdata vector to cache sectors ?
* olecftools: check UTF16 print support
* implement
  - libolecf_file_get_number_of_unallocated_blocks
  - libolecf_file_get_unallocated_block

* implement stream IO
* implement (document) summary information stream support
  implement property set stream support

* implement Workbook stream support
  in separate library
* implement olecf msg file support (Outlook MSG file format)
  in separate library

TEST
* add automated tests
* files with multi sector MSAT

API
* file io handle functions
* file information functions
* stream support
* allocation data support
* compound object support
* property set support

Library
* clone file io handle in item
* directory entry
  - debug print node color description
* property set stream
  - implement api functions and internal structures
  - debug print remnant and trailing data
  - debug print property names and values
* object stream
  - implement api functions and internal structures
* code clean up
  - check for code where list functions should be used instead
  - check for code where tree functions should be used instead

TOOLS
olecfinfo
* print basic file data (partial)
* print allocation data
* print stream/directory data (partial)
* print compobj stream info

20140930
* see `git log' for more recent change log
* removed README.macosx
* changes for project site move

20140810
* worked on libfwps support

20140801
* bug fix in Python-bindings

20140727
* worked on olecfmount Dokan support

20140726
* updated dependencies

20140329
* small changes to Makefile

20140323
* worked on Python bindings

20140317
* worked on setup.py

20140301
* 2014 update
* updated dependencies
* worked on Python bindings

20131209
* updated dependencies

20131108
* added support for 0 value sub directory entry not being used as index.

20131012
* added format version functions

20131011
* changes for non 512 sector sized files

20131009
* changes to dpkg files

20131008
* improved pyolecf as integer functions

20130929
* worked on setup.py, largely for MSI builds

20120928
* updated libfvalue and libfdatetime for filetime to integer functionality

20120925
* worked on API

20120924
* worked on pyolecf
* updated msvscpp files

20120923
* code clean up
* fixed memory leak in olecfexport for 0 byte streams

20120918
* worked on API
* olecfinfo: worked on property values

20120917
* code clean up
* worked on tests
* olecfmount: added extended options
* worked on API
* olecfinfo: worked on property values

20120916
* updated dependencies

20120717
* updated dependencies
* worked on tests
* olecfexport: fixed .export suffix

20120609
* updated dependencies

20120421
* worked on tests

20120414
* Textual changes
* worked on tests
* updates and bug fixes in pyolecf
* fixed codepage 949, 950 and 1255 restriction

20120413
* updated dependencies

20120319
* updated dependencies

20130305
* worked on libcdata integration
* worked on libfole

20130303
* worked on PackageMaker files

20130219
* worked on python bindings

20130218
* worked on python bindings

20130217
* fixed bug introduced in libcdata update
* worked on python bindings
* code clean up
* updated msvscpp files
* fixes in pc and spec file

20130211
* small changes

20130210
* 2013 update
* updated olecfmount added -X
* fixed olecfmount reading files
* worked on libcdata changes
* updated dependencies

20121104
* updated dependencies

20120920
* worked on libcdata changes

20120919
* updated dependencies

20120607
* olecfexport
 - worked on moving olecf file into export handle
* worked on using libfvalue for property values

20120530
* removed libcsystem_file_stream from export handle
  replaced by libcfile file
* code clean up remove item attach and detach functions

20120526
* fixes for olecfinfo requiring SummaryInformation and
  DocumentSummaryInformation directory entries
* updated dependencies
* worked on olecfmount
* added functions to retrieve item creation and modification time
* added missing libolecf_item_get_sub_item_by_utf16_name

20120523
* updated dependencies
* changes due to project move

20120508
* updated dependencies
* code clean up

20120503
* fixed issue in olecfexport regarding make directory
* applied updates

20120102
* 2012 update

20111023
* updated configure.ac and m4 scripts
* updated README files
* updated spec an pc file
* updated common, array type, libsystem, libuna, libbfio, libfdatetime,
  libfvalue

20110704
* worked on multi value support
* added libfvalue

20110703
* changes for: warning: AC_LANG_CONFTEST: no AC_LANG_SOURCE call detected in body
* updated libuna, libbfio, libfdatetime
* refactored property set reading code

olecftools/olecfmount.c
/*
 * Mounts an Object Linking and Embedding (OLE) Compound File (CF).
 *
 * Copyright (C) 2008-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "mount_dokan.h"
#include "mount_fuse.h"
#include "mount_handle.h"
#include "olecftools_getopt.h"
#include "olecftools_i18n.h"
#include "olecftools_libcerror.h"
#include "olecftools_libclocale.h"
#include "olecftools_libcnotify.h"
#include "olecftools_libolecf.h"
#include "olecftools_output.h"
#include "olecftools_signal.h"
#include "olecftools_unused.h"

mount_handle_t *olecfmount_mount_handle = NULL;
int olecfmount_abort                    = 0;

/* Signal handler for olecfmount
 */
void olecfmount_signal_handler(
      olecftools_signal_t signal OLECFTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "olecfmount_signal_handler";

	OLECFTOOLS_UNREFERENCED_PARAMETER( signal )

	olecfmount_abort = 1;

	if( olecfmount_mount_handle != NULL )
	{
		if( mount_handle_signal_abort(
		     olecfmount_mount_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal mount handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use olecfmount to mount an Object Linking and Embedding (OLE) Compound File (CF).";

	olecftools_option_t options[ ] = {
		{ 'c', "codepage", "codepage of ASCII strings, options: ascii, windows-874, windows-932, windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 (default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 or windows-1258" },
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr, while olecfmount will remain running in the foreground" },
		{ 'V', NULL, "print version" },
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
		{ 'X', "extended_options", "extended options to pass to sub system" },
#endif
		{ 0, "file", "an Object Linking and Embedding (OLE) Compound File (CF)" },
		{ 0, "mount_point", "the directory to serve as mount point" },
	};
	system_character_t options_string[ 32 ];

	libolecf_error_t *error                     = NULL;
	system_character_t *option_codepage         = NULL;
	system_character_t *source                  = NULL;
	char *program                               = "olecfmount";
	system_integer_t option                     = 0;
	int number_of_options                       = (int) ( sizeof( options ) / sizeof( olecftools_option_t ) );
	int result                                  = 0;
	int verbose                                 = 0;

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	system_character_t *mount_point             = NULL;
#endif

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	struct fuse_operations olecfmount_fuse_operations;

	system_character_t *option_extended_options = NULL;

#if defined( HAVE_LIBFUSE3 )
	/* Need to set this to 1 even if there no arguments, otherwise this causes
	 * fuse: empty argv passed to fuse_session_new()
	 */
	char *fuse_argv[ 2 ]                         = { program, NULL };
	struct fuse_args olecfmount_fuse_arguments   = FUSE_ARGS_INIT(1, fuse_argv);
#else
	struct fuse_args olecfmount_fuse_arguments   = FUSE_ARGS_INIT(0, NULL);
	struct fuse_chan *olecfmount_fuse_channel    = NULL;
#endif
	struct fuse *olecfmount_fuse_handle          = NULL;

#elif defined( HAVE_LIBDOKAN )
	DOKAN_OPERATIONS olecfmount_dokan_operations;
	DOKAN_OPTIONS olecfmount_dokan_options;
#endif

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "olecftools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( olecftools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	olecftools_output_version_fprint(
	 stdout,
	 program );

	if( olecftools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = olecftools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				olecftools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'c':
				option_codepage = optarg;

				break;

			case (system_integer_t) 'h':
				olecftools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				olecftools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
			case (system_integer_t) 'X':
				option_extended_options = optarg;

				break;
#endif
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		olecftools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind++ ];

	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing mount point.\n" );

		olecftools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	mount_point = argv[ optind ];
#endif
	libcnotify_verbose_set(
	 verbose );
	libolecf_notify_set_stream(
	 stderr,
	 NULL );
	libolecf_notify_set_verbose(
	 verbose );

	if( mount_handle_initialize(
	     &olecfmount_mount_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize mount handle.\n" );

		goto on_error;
	}
	if( option_codepage != NULL )
	{
		result = mount_handle_set_ascii_codepage(
		          olecfmount_mount_handle,
		          option_codepage,
		          &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set ASCII codepage in mount handle.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported ASCII codepage defaulting to: windows-1252.\n" );
		}
	}
	if( mount_handle_open(
	     olecfmount_mount_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source file\n" );

		goto on_error;
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( option_extended_options != NULL )
	{
#if defined( HAVE_LIBFUSE3 )
		// fuse_opt_add_arg: Assertion `!args->argv || args->allocated' failed.
		olecfmount_fuse_arguments.argc = 0;
		olecfmount_fuse_arguments.argv = NULL;
#endif
		/* This argument is required but ignored
		 */
		if( fuse_opt_add_arg(
		     &olecfmount_fuse_arguments,
		     "" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &olecfmount_fuse_arguments,
		     "-o" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &olecfmount_fuse_arguments,
		     option_extended_options ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
	}
	if( memory_set(
	     &olecfmount_fuse_operations,
	     0,
	     sizeof( struct fuse_operations ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear fuse operations.\n" );

		goto on_error;
	}
	olecfmount_fuse_operations.open       = &mount_fuse_open;
	olecfmount_fuse_operations.read       = &mount_fuse_read;
	olecfmount_fuse_operations.release    = &mount_fuse_release;
	olecfmount_fuse_operations.opendir    = &mount_fuse_opendir;
	olecfmount_fuse_operations.readdir    = &mount_fuse_readdir;
	olecfmount_fuse_operations.releasedir = &mount_fuse_releasedir;
	olecfmount_fuse_operations.getattr    = &mount_fuse_getattr;
	olecfmount_fuse_operations.destroy    = &mount_fuse_destroy;

#if defined( HAVE_LIBFUSE3 )
	olecfmount_fuse_handle = fuse_new(
	                          &olecfmount_fuse_arguments,
	                          &olecfmount_fuse_operations,
	                          sizeof( struct fuse_operations ),
	                          olecfmount_mount_handle );

	if( olecfmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
	result = fuse_mount(
	          olecfmount_fuse_handle,
	          mount_point );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to fuse mount file system.\n" );

		goto on_error;
	}
#else
	olecfmount_fuse_channel = fuse_mount(
	                           mount_point,
	                           &olecfmount_fuse_arguments );

	if( olecfmount_fuse_channel == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse channel.\n" );

		goto on_error;
	}
	olecfmount_fuse_handle = fuse_new(
	                          olecfmount_fuse_channel,
	                          &olecfmount_fuse_arguments,
	                          &olecfmount_fuse_operations,
	                          sizeof( struct fuse_operations ),
	                          olecfmount_mount_handle );

	if( olecfmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
#endif /* defined( HAVE_LIBFUSE3 ) */

	if( verbose == 0 )
	{
		if( fuse_daemonize(
		     0 ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable to daemonize fuse.\n" );

			goto on_error;
		}
	}
	result = fuse_loop(
	          olecfmount_fuse_handle );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to run fuse loop.\n" );

		goto on_error;
	}
#if defined( HAVE_LIBFUSE3 )
	fuse_unmount(
	 olecfmount_fuse_handle );
#else
	fuse_unmount(
	 mount_point,
	 olecfmount_fuse_channel );
#endif
	fuse_destroy(
	 olecfmount_fuse_handle );

	fuse_opt_free_args(
	 &olecfmount_fuse_arguments );

	return( EXIT_SUCCESS );

#elif defined( HAVE_LIBDOKAN )
	if( memory_set(
	     &olecfmount_dokan_operations,
	     0,
	     sizeof( DOKAN_OPERATIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan operations.\n" );

		goto on_error;
	}
	if( memory_set(
	     &olecfmount_dokan_options,
	     0,
	     sizeof( DOKAN_OPTIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan options.\n" );

		goto on_error;
	}
	olecfmount_dokan_options.Version    = DOKAN_VERSION;
	olecfmount_dokan_options.MountPoint = mount_point;

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	olecfmount_dokan_options.SingleThread = TRUE;
#else
	olecfmount_dokan_options.ThreadCount  = 0;
#endif
	if( verbose != 0 )
	{
		olecfmount_dokan_options.Options |= DOKAN_OPTION_STDERR;
#if defined( HAVE_DEBUG_OUTPUT )
		olecfmount_dokan_options.Options |= DOKAN_OPTION_DEBUG;
#endif
	}
/* This will only affect the drive properties
	olecfmount_dokan_options.Options |= DOKAN_OPTION_REMOVABLE;
*/

#if ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 )
	olecfmount_dokan_options.Options |= DOKAN_OPTION_KEEP_ALIVE;

	olecfmount_dokan_operations.CreateFile           = &mount_dokan_CreateFile;
	olecfmount_dokan_operations.OpenDirectory        = &mount_dokan_OpenDirectory;
	olecfmount_dokan_operations.CreateDirectory      = NULL;
	olecfmount_dokan_operations.Cleanup              = NULL;
	olecfmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	olecfmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	olecfmount_dokan_operations.WriteFile            = NULL;
	olecfmount_dokan_operations.FlushFileBuffers     = NULL;
	olecfmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	olecfmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	olecfmount_dokan_operations.FindFilesWithPattern = NULL;
	olecfmount_dokan_operations.SetFileAttributes    = NULL;
	olecfmount_dokan_operations.SetFileTime          = NULL;
	olecfmount_dokan_operations.DeleteFile           = NULL;
	olecfmount_dokan_operations.DeleteDirectory      = NULL;
	olecfmount_dokan_operations.MoveFile             = NULL;
	olecfmount_dokan_operations.SetEndOfFile         = NULL;
	olecfmount_dokan_operations.SetAllocationSize    = NULL;
	olecfmount_dokan_operations.LockFile             = NULL;
	olecfmount_dokan_operations.UnlockFile           = NULL;
	olecfmount_dokan_operations.GetFileSecurity      = NULL;
	olecfmount_dokan_operations.SetFileSecurity      = NULL;
	olecfmount_dokan_operations.GetDiskFreeSpace     = NULL;
	olecfmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	olecfmount_dokan_operations.Unmount              = &mount_dokan_Unmount;

#else
	olecfmount_dokan_operations.ZwCreateFile         = &mount_dokan_ZwCreateFile;
	olecfmount_dokan_operations.Cleanup              = NULL;
	olecfmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	olecfmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	olecfmount_dokan_operations.WriteFile            = NULL;
	olecfmount_dokan_operations.FlushFileBuffers     = NULL;
	olecfmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	olecfmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	olecfmount_dokan_operations.FindFilesWithPattern = NULL;
	olecfmount_dokan_operations.SetFileAttributes    = NULL;
	olecfmount_dokan_operations.SetFileTime          = NULL;
	olecfmount_dokan_operations.DeleteFile           = NULL;
	olecfmount_dokan_operations.DeleteDirectory      = NULL;
	olecfmount_dokan_operations.MoveFile             = NULL;
	olecfmount_dokan_operations.SetEndOfFile         = NULL;
	olecfmount_dokan_operations.SetAllocationSize    = NULL;
	olecfmount_dokan_operations.LockFile             = NULL;
	olecfmount_dokan_operations.UnlockFile           = NULL;
	olecfmount_dokan_operations.GetFileSecurity      = NULL;
	olecfmount_dokan_operations.SetFileSecurity      = NULL;
	olecfmount_dokan_operations.GetDiskFreeSpace     = NULL;
	olecfmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	olecfmount_dokan_operations.Unmounted            = NULL;
	olecfmount_dokan_operations.FindStreams          = NULL;
	olecfmount_dokan_operations.Mounted              = NULL;

#endif /* ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 ) */

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanInit();
#endif
	result = DokanMain(
	          &olecfmount_dokan_options,
	          &olecfmount_dokan_operations );

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanShutdown();
#endif
	switch( result )
	{
		case DOKAN_SUCCESS:
			break;

		case DOKAN_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: generic error\n" );
			break;

		case DOKAN_DRIVE_LETTER_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: bad drive letter\n" );
			break;

		case DOKAN_DRIVER_INSTALL_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to load driver\n" );
			break;

		case DOKAN_START_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: driver error\n" );
			break;

		case DOKAN_MOUNT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to assign drive letter\n" );
			break;

		case DOKAN_MOUNT_POINT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: mount point error\n" );
			break;

		default:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unknown error: %d\n",
			 result );
			break;
	}
	return( EXIT_SUCCESS );

#else
	fprintf(
	 stderr,
	 "No sub system to mount Object Linking and Embedding (OLE) Compound File (CF) format.\n" );

	return( EXIT_FAILURE );

#endif /* defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) */

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( olecfmount_fuse_handle != NULL )
	{
		fuse_destroy(
		 olecfmount_fuse_handle );
	}
	fuse_opt_free_args(
	 &olecfmount_fuse_arguments );
#endif
	if( olecfmount_mount_handle != NULL )
	{
		mount_handle_free(
		 &olecfmount_mount_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

